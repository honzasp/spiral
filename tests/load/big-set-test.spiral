(program
  (import std)
  (import std.test)
  (import std.set)

  (test "sieve of eratosthenes" (lambda (t)
    (var limit 100000)
    (fun make-sieve (i sieve)
      (if (<= i limit)
        (make-sieve (+ i 1) (set-insert sieve i))
        sieve))
    (fun remove-multiples (p i sieve)
      (if (<= (* p i) limit)
        (remove-multiples p (+ i 1) (set-remove sieve (* p i)))
        sieve))
    (fun walk-through-sieve (p sieve)
      (cond ((> p limit) sieve)
            ((set-contains? sieve p)
              (walk-through-sieve (+ p 1) (remove-multiples p 2 sieve)))
            (true
              (walk-through-sieve (+ p 1) sieve))))

    (let ((primes (walk-through-sieve 2 (make-sieve 2 (set-new -)))))
      (assert-eqv t (set-len primes) 9592)
      (assert-eqv t (set-get-index primes 0) 2)
      (assert-eqv t (set-get-index primes 1) 3)
      (assert-eqv t (set-get-index primes 2) 5)
      (assert-eqv t (set-get-index primes 3) 7)
      (assert-eqv t (set-get-index primes 4) 11)
      (assert-eqv t (set-get-index primes 99) 541)
      (assert-eqv t (set-get-index primes 999) 7919)
      (assert-eqv t (set-get-index primes 7999) 81799)
      (assert-eqv t (set-find-index primes 683) 123)
      (assert-eqv t (set-find-index primes 1567) 246)
      (assert-eqv t (set-find-index primes 77899) 7654)))))
