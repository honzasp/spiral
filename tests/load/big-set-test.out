test sieve of eratosthenes ok
